src/accelerate/utils/modeling.py

```diff
@@ -645,31 +711,20 @@ def find_tied_parameters(model: nn.Module, **kwargs):
     [['linear1.weight', 'linear2.weight']]
     ```
     """
-    # Initialize result and named_parameters before recursing.
-    named_parameters = kwargs.get("named_parameters", None)
-    prefix = kwargs.get("prefix", "")
-    result = kwargs.get("result", {})
 
-    if named_parameters is None:
-        named_parameters = {n: p for n, p in model.named_parameters()}
-    else:
-        # A tied parameter will not be in the full `named_parameters` seen above but will be in the `named_parameters`
-        # of the submodule it belongs to. So while recursing we track the names that are not in the initial
-        # `named_parameters`.
-        for name, parameter in model.named_parameters():
-            full_name = name if prefix == "" else f"{prefix}.{name}"
-            if full_name not in named_parameters:
-                # When we find one, it has to be one of the existing parameters.
-                for new_name, new_param in named_parameters.items():
-                    if new_param is parameter:
-                        if new_name not in result:
-                            result[new_name] = []
-                        result[new_name].append(full_name)
-
-    # Once we have treated direct parameters, we move to the child modules.
-    for name, child in model.named_children():
-        child_name = name if prefix == "" else f"{prefix}.{name}"
-        find_tied_parameters(child, named_parameters=named_parameters, prefix=child_name, result=result)
+    all_named_parameters = {name: param for name, param in _get_named_parameters(model, remove_duplicate=False)}
+    no_duplicate_named_parameters = {
+        name: param for name, param in _get_named_parameters(model, remove_duplicate=True)
+    }
+    diff_keys = set(all_named_parameters.keys()) - set(no_duplicate_named_parameters.keys())
+
+    result = {}
+    for key in diff_keys:
+        for name, param in no_duplicate_named_parameters.items():
+            if param is all_named_parameters[key]:
+                if name not in result:
+                    result[name] = []
+                result[name].append(key)
```

Comment:
Could you add a few lines of comment just like in the previous version to better understand how we find the tied weights using no_duplicate_named_parameters and all_named_parameters ?

Reply:
Nice ! thanks a lot !